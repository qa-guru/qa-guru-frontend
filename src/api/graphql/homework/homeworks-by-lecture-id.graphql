query homeWorksByLectureId(
  $offset: Int!
  $limit: Int!
  $sort: StudentHomeWorkSort!
  $lectureId: ID!
) {
  homeWorksByLectureId(
    offset: $offset
    limit: $limit
    sort: $sort
    lectureId: $lectureId
  ) {
    items {
      id
      answer
      status
      student {
        id
        firstName
        middleName
        avatar
        lastName
        rating {
          rating
        }
      }
      mentor {
        id
        firstName
        middleName
        lastName
        avatar
        rating {
          rating
        }
      }
      creationDate
      startCheckingDate
      endCheckingDate
    }
    offset
    limit
    totalElements
  }
}
