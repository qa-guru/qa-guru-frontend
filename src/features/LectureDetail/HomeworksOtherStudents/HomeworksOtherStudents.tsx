import React, { useEffect, useState } from "react";
import {
  Button,
  Dialog,
  DialogContent,
  Paper,
  Stack,
  Typography,
} from "@mui/material";
import { LoadingButton } from "@mui/lab";
import ExpandMoreIcon from "@mui/icons-material/ExpandMore";
import { useModal } from "react-modal-hook";
import { IHomeworksOtherStudents } from "./HomeworksOtherStudents.types";
import { primary } from "../../../theme/colors";
import { HomeworkItem } from "../HomeworkItem";
import Comment from "../Comment/Comment";
import CommentContainer from "../Comment";

const style = {
  paper: { padding: { xs: "15px", md: "20px" } },
  loadMoreBtn: { color: primary.main, margin: "0 auto" },
  scrollContainer: { overflowY: "auto", maxHeight: "calc(100vh - 200px)" },
};

const ModalWrapper: React.FC<{ item: any; userId: string }> = ({
  item,
  userId,
}) => {
  const [showModal, hideModal] = useModal(({ in: open }) => (
    <Dialog open={open} onClose={hideModal} maxWidth="xl" fullWidth>
      <DialogContent sx={style.scrollContainer} id="scroll-container">
        <HomeworkItem dataHomeWorkByLecture={item} dataUserId={userId} />
        <CommentContainer id={item.id}>
          <Comment />
        </CommentContainer>
      </DialogContent>
    </Dialog>
  ));

  return (
    <Button sx={{ mt: "10px" }} variant="contained" onClick={showModal}>
      Показать
    </Button>
  );
};

const HomeworksOtherStudents: React.FC<IHomeworksOtherStudents> = (props) => {
  const { data, fetchMore, dataUserId } = props;
  const { id: userId } = dataUserId.user!;
  const { items, offset, totalElements, limit } = data.homeWorksByLectureId!;
  const [loading, setLoading] = useState<boolean>(false);
  const [hasMoreHomeworks, setHasMoreHomeworks] = useState<boolean>(true);

  useEffect(() => {
    if (items?.length! >= totalElements) {
      setHasMoreHomeworks(false);
    }
  }, [items]);

  const handleLoadMore = () => {
    setLoading(true);
    fetchMore({
      variables: {
        offset: items?.length,
      },
      updateQuery: (
        prev: { homeWorksByLectureId: { items: any } },
        { fetchMoreResult }: any
      ) => {
        if (!fetchMoreResult) return prev;
        return {
          homeWorksByLectureId: {
            ...fetchMoreResult.homeWorksByLectureId,
            items: [
              ...prev.homeWorksByLectureId.items,
              ...fetchMoreResult.homeWorksByLectureId.items,
            ],
          },
        };
      },
    }).then(() => {
      setLoading(false);
    });
  };

  return (
    <>
      <Stack mb="15px" pt="30px" spacing={1} direction="row">
        <Typography variant="h4">Домашние работы других студентов</Typography>
        <Typography variant="h4">({totalElements})</Typography>
      </Stack>
      <Stack spacing={2}>
        {items?.map((item, index) => {
          return (
            <Paper key={index} sx={style.paper}>
              <HomeworkItem
                dataHomeWorkByLecture={item!}
                dataUserId={userId!}
              />
              <ModalWrapper
                key={`modal-${index}`}
                item={item}
                userId={userId!}
              />
            </Paper>
          );
        })}
      </Stack>
      {hasMoreHomeworks && (
        <Stack mt="15px">
          <LoadingButton
            loading={loading}
            onClick={handleLoadMore}
            sx={style.loadMoreBtn}
            endIcon={<ExpandMoreIcon />}
          >
            Загрузить еще
          </LoadingButton>
        </Stack>
      )}
    </>
  );
};

export default HomeworksOtherStudents;
