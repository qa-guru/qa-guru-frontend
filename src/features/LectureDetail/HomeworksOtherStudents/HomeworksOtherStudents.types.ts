import { HomeWorksByLectureIdQuery } from "../../../api/graphql/generated/graphql";

export interface IHomeworksOtherStudents {
  data: HomeWorksByLectureIdQuery;
}
