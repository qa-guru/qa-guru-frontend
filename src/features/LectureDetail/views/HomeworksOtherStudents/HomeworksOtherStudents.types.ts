import {
  HomeWorksByLectureIdQuery,
  UserIdQuery,
} from "../../../../api/graphql/generated/graphql";

export interface IHomeworksOtherStudents {
  data: HomeWorksByLectureIdQuery;
  fetchMore: (params: {
    variables: { offset?: number };
    updateQuery: (
      prev: HomeWorksByLectureIdQuery,
      { fetchMoreResult }: { fetchMoreResult?: HomeWorksByLectureIdQuery | undefined }
    ) => HomeWorksByLectureIdQuery;
  }) => Promise<void>;  dataUserId: UserIdQuery;
}
