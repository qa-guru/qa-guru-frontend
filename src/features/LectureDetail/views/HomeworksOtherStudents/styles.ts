import { primary } from "../../../../theme/colors";

export const style = {
  paper: { padding: { xs: "15px", md: "20px" } },
  loadMoreBtn: { color: primary.main, margin: "0 auto" },
};
