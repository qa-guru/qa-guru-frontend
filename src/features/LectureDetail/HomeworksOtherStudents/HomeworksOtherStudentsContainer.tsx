import React from "react";
import { useParams } from "react-router-dom";
import HomeworksOtherStudents from "./HomeworksOtherStudents";
import { useHomeWorksByLectureIdQuery } from "../../../api/graphql/homework/homeWorksByLectureId";
import {
  InputMaybe,
  Order,
  StudentHomeWorkSortField,
} from "../../../api/graphql/generated/graphql";
import NoDataErrorMessage from "../../../shared/NoDataErrorMessage";
import { useUserIdQuery } from "../../../api/graphql/user/userId";
import Spinner from "../../../shared/Spinner";

const HomeworksOtherStudentsContainer: React.FC = () => {
  const { lectureId } = useParams();
  const fieldSortHomeworks =
    "CREATION_DATE" as InputMaybe<StudentHomeWorkSortField>;
  const fieldOrderHomeworks = "DESC" as InputMaybe<Order>;

  const { data: dataUserId, loading: loadingUserId } = useUserIdQuery();
  const { data, loading, fetchMore } = useHomeWorksByLectureIdQuery({
    variables: {
      offset: 0,
      limit: 2,
      sort: {
        field: fieldSortHomeworks,
        order: fieldOrderHomeworks,
      },
      lectureId: lectureId!,
    },
  });

  if (loading || loadingUserId) return <Spinner />;
  if (!data || !dataUserId) return <NoDataErrorMessage />;

  return (
    <HomeworksOtherStudents
      dataUserId={dataUserId}
      fetchMore={fetchMore}
      data={data}
    />
  );
};

export default HomeworksOtherStudentsContainer;
