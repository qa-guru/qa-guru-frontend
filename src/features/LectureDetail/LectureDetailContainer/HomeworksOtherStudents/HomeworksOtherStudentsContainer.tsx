import React from "react";
import { useParams } from "react-router-dom";
import {
  InputMaybe,
  Order,
  StudentHomeWorkSortField,
} from "../../../../api/graphql/generated/graphql";
import { useUserIdQuery } from "../../../../api/graphql/user/userId";
import HomeworksOtherStudents from "../../LectureDetailView/components/HomeworksOtherStudents";
import NoDataErrorMessage from "../../../../shared/NoDataErrorMessage";
import Spinner from "../../../../shared/Spinner";
import { useHomeWorksByLectureIdQuery } from "../../../../api/graphql/homework/homeWorksByLectureId";

const HomeworksOtherStudentsContainer: React.FC = () => {
  const { lectureId } = useParams();

  const sortOptions = {
    field: "CREATION_DATE" as InputMaybe<StudentHomeWorkSortField>,
    order: "DESC" as InputMaybe<Order>,
  };

  const { data: dataUserId, loading: loadingUserId } = useUserIdQuery();
  const { data, loading, fetchMore } = useHomeWorksByLectureIdQuery({
    variables: {
      offset: 0,
      limit: 2,
      sort: sortOptions,
      lectureId: lectureId!,
    },
  });

  if (loading || loadingUserId) return <Spinner />;
  if (!data || !dataUserId) return <NoDataErrorMessage />;

  return (
    <HomeworksOtherStudents
      dataUserId={dataUserId}
      fetchMore={fetchMore}
      data={data}
    />
  );
};

export default HomeworksOtherStudentsContainer;
