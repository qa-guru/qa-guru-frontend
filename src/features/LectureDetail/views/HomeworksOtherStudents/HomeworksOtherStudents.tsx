import React, { useEffect, useState } from "react";
import { Paper, Stack, Typography } from "@mui/material";
import { LoadingButton } from "@mui/lab";
import ExpandMoreIcon from "@mui/icons-material/ExpandMore";
import { IHomeworksOtherStudents } from "./HomeworksOtherStudents.types";
import { style } from "./styles";
import ModalHomeworksOtherStudents from "../ModalHomeworksOtherStudents";
import HomeworkItem from "../HomeworkItem";

const HomeworksOtherStudents: React.FC<IHomeworksOtherStudents> = (props) => {
  const { data, fetchMore, dataUserId } = props;
  const { items, offset, totalElements, limit } = data.homeWorksByLectureId!;
  const [loading, setLoading] = useState<boolean>(false);
  const [hasMoreHomeworks, setHasMoreHomeworks] = useState<boolean>(true);

  useEffect(() => {
    if (items?.length! >= totalElements) {
      setHasMoreHomeworks(false);
    }
  }, [items]);

  const handleLoadMore = () => {
    setLoading(true);
    fetchMore({
      variables: {
        offset: items?.length,
      },
      updateQuery: (
        prev: { homeWorksByLectureId: { items: any } },
        { fetchMoreResult }: any
      ) => {
        if (!fetchMoreResult) return prev;
        return {
          homeWorksByLectureId: {
            ...fetchMoreResult.homeWorksByLectureId,
            items: [
              ...prev.homeWorksByLectureId.items,
              ...fetchMoreResult.homeWorksByLectureId.items,
            ],
          },
        };
      },
    }).then(() => {
      setLoading(false);
    });
  };

  return (
    <>
      <Stack my={3} pt={3} spacing={1} direction="row">
        <Typography variant="h4">Домашние работы других студентов</Typography>
        <Typography variant="h4">({totalElements})</Typography>
      </Stack>
      <Stack spacing={2}>
        {items?.map((item, index) => {
          return (
            <Paper key={index} sx={style.paper}>
              <HomeworkItem
                dataHomeWorkByLecture={item!}
                dataUserId={dataUserId}
              />
              <ModalHomeworksOtherStudents
                key={index}
                item={item!}
                dataUserId={dataUserId}
              />
            </Paper>
          );
        })}
      </Stack>
      {hasMoreHomeworks && (
        <Stack mt="15px">
          <LoadingButton
            loading={loading}
            onClick={handleLoadMore}
            sx={style.loadMoreBtn}
            endIcon={<ExpandMoreIcon />}
          >
            Загрузить еще
          </LoadingButton>
        </Stack>
      )}
    </>
  );
};

export default HomeworksOtherStudents;
