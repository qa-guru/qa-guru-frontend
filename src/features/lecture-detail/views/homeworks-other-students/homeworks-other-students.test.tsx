import { MemoryRouter } from "react-router-dom";
import HomeworksOtherStudents from "./homeworks-other-students";
import { homeWorksByLectureId } from "../../../../shared/mocks/homeworks-by-lecture-id.mock";
import { render } from "../../../../test/utils-test";
import { userId } from "../../../../shared/mocks/user-id.mock";

const fetchMore: any = () => {};

describe("HomeworksOtherStudents", () => {
  it("the component is render", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <HomeworksOtherStudents
          dataUserId={userId}
          fetchMore={fetchMore}
          data={homeWorksByLectureId}
        />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
