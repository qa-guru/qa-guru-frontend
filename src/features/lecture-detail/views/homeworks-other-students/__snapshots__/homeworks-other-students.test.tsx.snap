// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HomeworksOtherStudents > renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiStack-root css-1t3znw4-MuiStack-root"
  >
    <h4
      class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
    >
      Домашние работы других студентов
    </h4>
    <h4
      class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
    >
      (7)
    </h4>
  </div>
  <div
    class="MuiStack-root css-1sazv7p-MuiStack-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-nxklxt-MuiPaper-root"
    >
      <div
        class="MuiStack-root css-osjy2d-MuiStack-root"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
        >
          Ответ на задание
        </h5>
        <div
          class="MuiStack-root css-kdc3n8-MuiStack-root"
        >
          <svg
            fill="none"
            height="14"
            viewBox="0 0 18 14"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6.00016 11.17L1.83016 7L0.410156 8.41L6.00016 14L18.0002 2L16.5902 0.59L6.00016 11.17Z"
              fill="#2CCCA6"
            />
          </svg>
          <div
            class="MuiBox-root css-0"
          >
            <p
              class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
            >
              Принято
            </p>
          </div>
        </div>
        <div
          class="MuiStack-root css-vq7ph6-MuiStack-root"
        >
          <div
            class="MuiBox-root css-1quw33y"
          >
            <div
              class="MuiStack-root css-qix97k-MuiStack-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-11h8gky-MuiAvatar-root"
              >
                <h6
                  class="MuiTypography-root MuiTypography-subtitle2 css-ati914-MuiTypography-root"
                >
                  ss
                </h6>
              </div>
              <div
                class="MuiBox-root css-0"
              >
                <h6
                  class="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
                >
                  string string
                </h6>
                <div
                  class="MuiStack-root css-gmwslw-MuiStack-root"
                >
                  <h6
                    class="MuiTypography-root MuiTypography-subtitle2 css-ati914-MuiTypography-root"
                  >
                    23.05.2023 | 00:00
                  </h6>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      <div
        class="MuiBox-root css-1yuhvjn"
      >
        <div
          class="MuiStack-root css-qix97k-MuiStack-root"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-11h8gky-MuiAvatar-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 css-ati914-MuiTypography-root"
            >
              ss
            </h6>
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
            >
              string string
            </h6>
            <div
              class="MuiStack-root css-gmwslw-MuiStack-root"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle2 css-ati914-MuiTypography-root"
              >
                23.05.2023 | 00:00
              </h6>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root css-6qfr2n"
      >
        <div
          class="MuiBox-root css-i6bazn"
        >
          <h6
            class="MuiTypography-root MuiTypography-subtitle1 css-18u75fl-MuiTypography-root"
          >
            string
            <br />
          </h6>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1ouw2p3-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Редактировать
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <div
        class="MuiStack-root css-47fnhy-MuiStack-root"
      >
        <div
          class="MuiStack-root css-e8agqp-MuiStack-root"
        >
          <div
            class="MuiStack-root css-jfdv4h-MuiStack-root"
          >
            <div
              class="MuiStack-root css-kdc3n8-MuiStack-root"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-circular MuiSkeleton-pulse css-wkdwxb-MuiSkeleton-root"
                style="width: 40px; height: 40px;"
              />
              <div
                class="MuiStack-root css-1r50ci7-MuiStack-root"
              >
                <span
                  class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-12j6v5j-MuiSkeleton-root"
                />
                <span
                  class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-12j6v5j-MuiSkeleton-root"
                />
              </div>
            </div>
            <span
              class="MuiSkeleton-root MuiSkeleton-rounded MuiSkeleton-pulse css-7uzo9y-MuiSkeleton-root"
              style="width: 100%; height: 60px;"
            />
          </div>
          <div
            class="MuiStack-root css-jfdv4h-MuiStack-root"
          >
            <div
              class="MuiStack-root css-kdc3n8-MuiStack-root"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-circular MuiSkeleton-pulse css-wkdwxb-MuiSkeleton-root"
                style="width: 40px; height: 40px;"
              />
              <div
                class="MuiStack-root css-1r50ci7-MuiStack-root"
              >
                <span
                  class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-12j6v5j-MuiSkeleton-root"
                />
                <span
                  class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-12j6v5j-MuiSkeleton-root"
                />
              </div>
            </div>
            <span
              class="MuiSkeleton-root MuiSkeleton-rounded MuiSkeleton-pulse css-7uzo9y-MuiSkeleton-root"
              style="width: 100%; height: 60px;"
            />
          </div>
        </div>
        <div
          class="MuiBox-root css-0"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Показать
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiStack-root css-1inmmz-MuiStack-root"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiLoadingButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-m0rdlc-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
      id=":r0:"
      tabindex="0"
      type="button"
    >
      Загрузить еще
      <span
        class="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ExpandMoreIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
